python-checks/src/main/java/org/sonar/python/checks/LocalVariableAndParameterNameConventionCheck.java

```diff
@@ -78,7 +83,42 @@ private void checkName(Symbol symbol, SubscriptionContext ctx) {
   }
 
   private static boolean isType(Symbol symbol) {
-    return symbol.usages().stream().map(Usage::tree).filter(Expression.class::isInstance).map(Expression.class::cast).anyMatch(e -> e.type().mustBeOrExtend("type"));
+    return isExtendingType(symbol) || isAssignedFromSpecialForm(symbol);
+  }
+
+  private static boolean isExtendingType(Symbol symbol) {
+    return symbol.usages().stream().map(Usage::tree).filter(Expression.class::isInstance).map(Expression.class::cast).anyMatch(e -> e.type().mustBeOrExtend("type")) ||
+      Objects.equals(symbol.annotatedTypeName(), "typing._SpecialForm");
+  }
+
+  private static boolean isAssignedFromSpecialForm(Symbol symbol) {
+    List<Tree> assignmentTrees = symbol.usages().stream().filter(u -> u.kind() == Usage.Kind.ASSIGNMENT_LHS).map(Usage::tree).toList();
+    for (Tree assignmentTree : assignmentTrees) {
+      Expression assignedValue = getAssignedValue(assignmentTree);
+      if (assignedValue == null) {
+        continue;
+      }
+      if (assignedValue.is(Tree.Kind.SUBSCRIPTION)) {
+        SubscriptionExpression subscriptionExpression = (SubscriptionExpression) assignedValue;
+        if (subscriptionExpression.object().is(Tree.Kind.NAME)) {
+          Symbol assignedSymbol = ((Name) subscriptionExpression.object()).symbol();
+          if (assignedSymbol != null && isExtendingType(assignedSymbol)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static Expression getAssignedValue(Tree assignmentTree) {
```

Comment:
_Nitpick:_ Here, (and in line 95), what you have is a name that is part of an assignment rather than the assignment itself. To reflect this, I'd suggest renaming to `assignedName` (and `assignedNames` respectively).